### ISSUE TITLE
Propose new API to declare different types of 'configurations'

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#9834]
The concept of _configuration_ (inherited from ivy) as a general way to work with different sets of dependencies within one project has grown over times and it now encapsulate different aspects of "working with dependencies". This became particularly clear when the `canBeConsumed` and `canBeResolved` flags were introduced on the configuration API to make it ready for _variant aware dependency management_.

This is well explained in the [Dependency Management with Gradle Part 1 – Fundamentals](https://gradle.github.io/webinar-dep-mgmt-part-1/#/) webinar. In particular the slide on [configurations](https://gradle.github.io/webinar-dep-mgmt-part-1/#/configurations) states that we have _3 kinds of configurations_:

1. "buckets" of dependencies
2. resolvable configurations
3. consumable configurations

We currently propose users to be aware of this separation and use it. We will also deprecate certain "wrong" usages in Gradle 6 - #9353. However, the API at the moment is awkward to use as [this](https://stackoverflow.com/questions/56711728/run-all-subprojects-tests-using-gradle/56714097#56714097) example shows (which still does not make a clear separation between "buckets" and "consumables", because that would make it even more verbose).

The goal of this issue is to propose a new API - possibly independent of the existing `configurations {}` block - that talks about the three concepts above rather than "configurations". If we can name these concepts clearer, we can also move away from the term "configuration" in documentation, which is so generic that it is often hard to explain and understand.

Implementation wise, the new API can initially delegate to the existing Configuration API but should eventually also help to improve the implementation of the API (currently the implementation `DefaultConfiguration` is huge and mixes a lot of concerns in one class).

Another aspect to improve is that we currently have a distinction between "primary" and "secondary" variants. A `canBeConsumed=true` configurations is itself a (primary) variant, but it can cary additional variants accessible through `configuration.outgoing.variants`. The specific handling during matching is, that only the primary variant are matched first. Meaning, the attributes of the secondary variants can only be variations of the primary one, for example with a different `artifactType` attribute (e.g. `jar` vs `classes`). During artifact selection ("second" matching step) these variants are accessible.

**Comment by ducrohet** [comment]
Could we also separate configurations needed to build the projects, and configurations needed for the build itself?

In the Android plugin there are a few tools/libraries that we download dynamically during execution time and we do this via custom configurations. However we see people do `configurations.all { ...}` to set custom resolution strategy and this impacts these internal configurations. 

We'd love to have a separate configuration container for the build itself. thanks!

**Comment by melix** [comment]
Yes, this is indeed something we need to think about. We could just have different configuration containers. But I also had the need for a full "detached resolver" with its own repositories.

**Comment by jjohannes** [comment]
Initial spec document: https://docs.google.com/document/d/1mhNP7TRCTeRHKRirVvFUM2aEDz1chM6be926vAJW9Ns

**Comment by martinda** [comment]
@jjohannes Can you please allow public comments on the document? Or at the very least, make it viewable? Thanks.

**Comment by jjohannes** [comment]
Thanks for your interest in this topic @martinda @ducrohet @lacasseio.

The document I linked is in an internal space and I cannot share it. It's also in a very early state. I linked it here so that we find it back as we have to focus on some other things first before we can get back to this. But it's next on the list. Once we continue the work, I'll make sure to share more information for feedback here.
